### PR TITLE
Update bn layer forwarding @open sesame 09/15 21:11

### PR DESCRIPTION
**Pull request by zhoonit**
**Changes proposed in this PR:**
- Update bn layer to more naive implementation
- Update broken bn layer tc

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

### CODE REVIEW

**Comment by zhoonit** [comment]
Axis should be handled in more appropriate fashion. Need some rework

**Comment by zhoonit** [comment]
Now works for the case axis is 3. Can be reviewed for now and additional patches will follow

File: nntrainer/include/bn_layer.h [hunk 1]
```diff
@@ -42,7 +42,11 @@ class BatchNormalizationLayer : public Layer {
   /**
    * @brief     Constructor of Batch Noramlization Layer
    */
-  BatchNormalizationLayer() : epsilon(0.001) { setType(LAYER_BN); };
+  BatchNormalizationLayer(float eps = 0.001, int axis = -1) :
+    epsilon(eps),
+    axis(-1) {
```
Comment (kparichay) [inline nntrainer/include/bn_layer.h:47]: axis(axis)?

File: nntrainer/src/bn_layer.cpp [hunk 2]
```diff
@@ -34,13 +34,23 @@ namespace nntrainer {
...
   output_dim = input_dim;
-  TensorDim dim = input_dim;
-  dim.batch(1);
+  TensorDim dim;
+
+  /// @note this logic cannot tell channel is actually 1 or it is just not used.
+  if (axis == -1)
+    axis = input_dim.channel() > 1 ? 1 : 3;
+
+  dim.setTensorDim(axis, input_dim.getTensorDim(axis));
```
Comment (kparichay) [inline nntrainer/src/bn_layer.cpp:48]: with batch?

File: nntrainer/src/bn_layer.cpp [hunk 3]
```diff
@@ -83,37 +93,31 @@ sharedConstTensor BatchNormalizationLayer::forwarding(sharedConstTensor in) {
...
-    Tensor cmu;
-
-    cmu = input.average(0);
+  Tensor deviation;
 
+  input = *in;
+  /// @todo change trainable #524
+  if (trainable) {
+    Tensor cmu = input.average(axes_to_reduce);
+    std::cerr << input;
```
Comment (kparichay) [inline nntrainer/src/bn_layer.cpp:102]: debug code in PR?

File: nntrainer/src/bn_layer.cpp [hunk 4]
```diff
@@ -83,37 +93,31 @@ sharedConstTensor BatchNormalizationLayer::forwarding(sharedConstTensor in) {
...
+    std::cerr << cmu;
     deviation = input.subtract(cmu);
 
-    this->cvar = deviation.chain()
-                   .multiply_i(deviation)
-                   .sum(0)
-                   .multiply_i(1.0f / input_dim.batch())
-                   .add_i(epsilon)
-                   .run();
+    cvar = deviation.multiply(deviation).average(axes_to_reduce).add(epsilon);
```
Comment (kparichay) [inline nntrainer/src/bn_layer.cpp:106]: add -> add_i?

File: nntrainer/src/bn_layer.cpp [hunk 5]
```diff
@@ -83,37 +93,31 @@ sharedConstTensor BatchNormalizationLayer::forwarding(sharedConstTensor in) {
...
+    float momentum = 0.99;
     mu.multiply_i(momentum);
     mu.add_i(cmu, 1 - momentum);
     var.multiply_i(momentum);
     var.add_i(cvar, 1 - momentum);
 
     this->x_normalized = deviation.divide(cvar.apply(sqrtFloat));
-
-    this->hidden = x_normalized.chain().multiply_i(gamma).add_i(beta).run();
+    this->hidden = x_normalized.multiply(gamma).add(beta);
```
Comment (kparichay) [inline nntrainer/src/bn_layer.cpp:116]: add -> add_i

File: nntrainer/src/bn_layer.cpp [hunk 6]
```diff
@@ -83,37 +93,31 @@ sharedConstTensor BatchNormalizationLayer::forwarding(sharedConstTensor in) {
...
   } else {
-    /// NYI
-    throw std::runtime_error("not_yet_implemented");
+    deviation = input.subtract(mu);
+    this->x_normalized = deviation.divide(var.apply(sqrtFloat));
```
Comment (kparichay) [inline nntrainer/src/bn_layer.cpp:119]: I wonder why store var than std given you have to apply sqrt everytime.

File: nntrainer/src/bn_layer.cpp [hunk 7]
```diff
@@ -83,37 +93,31 @@ sharedConstTensor BatchNormalizationLayer::forwarding(sharedConstTensor in) {
...
+    this->hidden = x_normalized.multiply(gamma).add(beta);
```
Comment (kparichay) [inline nntrainer/src/bn_layer.cpp:120]: add_i

File: test/input_gen/genInput.py [hunk 8]
```diff
@@ -361,16 +363,17 @@ def bn_tf(x, *, trainable=True, init_beta=gen_tensor, init_gamma=gen_tensor, axi
     bnlayer = tf.keras.layers.BatchNormalization(
         axis=axis,
         trainable=trainable,
-        momentum=1.0,
+        momentum=0.99,
```
Comment (kparichay) [inline test/input_gen/genInput.py:366]: Please use `0.9` for momentum for unittests. 
This increases the multiplier (0.01 to 0.1) which will make difference bigger (if any) and easier to identify

File: test/input_gen/genInput.py [hunk 9]
```diff
@@ -577,8 +578,11 @@ def gen_test_case_bn(input_shape, base_name, training=True):
                 base_name = "tc_fc_1")
 
 # Bn layer unit test case:
-    if target == "bn_1":
-        gen_test_case_bn(input_shape = [3, 1, 4, 5], base_name = "tc_bn_1")
+    if target == "bn_fc_1":
+        gen_test_case_bn(input_shape = [3, 1, 1, 12], base_name = "tc_bn_fc_1", axis=-1)
+
+    if target == "bn_conv_2":
+        gen_test_case_bn(input_shape = [3, 2, 4, 5], base_name = "tc_bn_conv_2", axis=1)
```
Comment (kparichay) [inline test/input_gen/genInput.py:586]: make a unittest with batch size 1 as well.

File: nntrainer/src/bn_layer.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/src/bn_layer.cpp:48]: The batch should be 1 for params isn't it?

File: nntrainer/src/bn_layer.cpp (hunk 6 above)
Comment (zhoonit) [reply nntrainer/src/bn_layer.cpp:119]: Var is smoothed out with momentum when we need to optimize. we could save/load std instead of var to optimizer more. Let's keep it var for now since other frameworks store var as well and it is easier to match.

File: nntrainer/src/bn_layer.cpp (hunk 3 above)
Comment (zhoonit) [reply nntrainer/src/bn_layer.cpp:102]: Removed

File: nntrainer/include/bn_layer.h (hunk 1 above)
Comment (zhoonit) [reply nntrainer/include/bn_layer.h:47]: Done

File: nntrainer/src/bn_layer.cpp (hunk 4 above)
Comment (zhoonit) [reply nntrainer/src/bn_layer.cpp:106]: Done

File: nntrainer/src/bn_layer.cpp (hunk 5 above)
Comment (zhoonit) [reply nntrainer/src/bn_layer.cpp:116]: Done

File: nntrainer/src/bn_layer.cpp (hunk 7 above)
Comment (zhoonit) [reply nntrainer/src/bn_layer.cpp:120]: Done

File: test/input_gen/genInput.py (hunk 8 above)
Comment (zhoonit) [reply test/input_gen/genInput.py:366]: Done

File: nntrainer/src/bn_layer.cpp (hunk 6 above)
Comment (kparichay) [reply nntrainer/src/bn_layer.cpp:121]: then lets add a todo here to shift to std in the evaluate stage to save computation.

File: nntrainer/src/bn_layer.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/src/bn_layer.cpp:48]: aah input_dim is now without batch.

File: nntrainer/src/bn_layer.cpp (hunk 6 above)
Comment (zhoonit) [reply nntrainer/src/bn_layer.cpp:121]: Done

File: test/input_gen/genInput.py (hunk 9 above)
Comment (zhoonit) [reply test/input_gen/genInput.py:586]: I'll do this in other PR as even this tc is not used as of now.

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/src/bn_layer.cpp [hunk 10]
```diff
@@ -83,37 +94,34 @@ sharedConstTensor BatchNormalizationLayer::forwarding(sharedConstTensor in) {
...
 
-    /// @todo replace momentum paramter
+    /// @todo replace momentum parameter to prop
     float momentum = 0.9;
     mu.multiply_i(momentum);
     mu.add_i(cmu, 1 - momentum);
     var.multiply_i(momentum);
     var.add_i(cvar, 1 - momentum);
 
+    std::cerr << cvar;
```
Comment (kparichay) [inline nntrainer/src/bn_layer.cpp:115]: is this intentional? if you want to keep logging messages for debugging in patches, let's stick to `ml_logd`.

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 11]
```diff
@@ -846,22 +847,52 @@ TEST_F(nntrainer_BatchNormalizationLayer, checkValidation_01_p) {
...
+  }
+
+  virtual void prepareLayer() {
+    setProperty("input_shape=2:4:5 | epsilon=0.001 | batch_size=3");
+    setOptimizer(nntrainer::OptType::sgd, "learning_rate=1");
+  }
+};
+
+TEST_F(nntrainer_BatchNormalizationLayer_Conv,
+       DISABLED_forward_backward_training_01_p) {
```
Comment (kparichay) [inline test/unittest/unittest_nntrainer_layers.cpp:884]: is this DISABLED intended?

File: test/input_gen/genInput.py (hunk 9 above)
Comment (kparichay) [reply test/input_gen/genInput.py:586]: ok.

File: nntrainer/src/bn_layer.cpp (hunk 10 above)
Comment (zhoonit) [reply nntrainer/src/bn_layer.cpp:115]: No it's mistake. sry about that.

File: test/unittest/unittest_nntrainer_layers.cpp (hunk 11 above)
Comment (zhoonit) [reply test/unittest/unittest_nntrainer_layers.cpp:884]: Yes for now.

Comment (kparichay) [reply test/unittest/unittest_nntrainer_layers.cpp:884]: cool.

**Review by kparichay — APPROVED** [review]
LGTM
